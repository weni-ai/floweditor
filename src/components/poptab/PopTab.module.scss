@import 'src/variables.module.scss';

.pop_wrapper {
  position: fixed;
  z-index: $z_pop_tabs;

  &.visible {
    .popped {
      opacity: 1;
      transition: right 0.1s ease-out; // cubic-bezier(0.175, 0.885, 0.32, 1.275), opacity 0.4s;
      box-shadow: 0px 0px 8px 0px #666;

      &.revisions {
        background: $unnnic-color-background-snow;
        outline: $unnnic-border-width-thinner solid $unnnic-color-neutral-cloudy;
        outline-offset: -$unnnic-border-width-thinner;
        border-radius: $unnnic-border-radius-md;
        box-shadow: none;

        .body {
          padding: $unnnic-spacing-inset-xs;
        }
      }
    }

    z-index: $z_pop_tabs +1;

    .tab_wrapper {
      .tab {
        left: 0px;

        &:hover {
          left: 0px;
        }
      }
    }
  }

  .tab_wrapper {
    .tab {
      position: absolute;
      left: -3.5 * $unnnic-font-size;
      padding-right: 0.75 * $unnnic-font-size + 10;
      border-top-left-radius: $unnnic-border-radius-md;
      border-bottom-left-radius: $unnnic-border-radius-md;
      cursor: pointer;
      width: 130px;
      transition: left 0.1s ease-out;
      transition-delay: 300ms;
      box-sizing: border-box;
      display: flex;
      align-items: center;

      &:hover {
        left: -120px;
        transition: left 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275);

        .label {
          opacity: 1;
        }
      }

      .icon {
        display: inline-block;
        padding: $unnnic-spacing-stack-nano 0.75 * $unnnic-font-size;
      }

      .label {
        width: 60px;
        display: inline-block;
        color: #fff;
        opacity: 0;
        transition: opacity 0.8s;
        line-height: 16px;
      }
    }
  }

  .popped {
    .header {
      color: #fff;
      height: 26px;

      .header_label {
        font-size: 13px;
        font-weight: 400;
        padding: 5px 10px;
      }

      .close {
        cursor: pointer;
        font-size: 16px;
        float: right;
        padding: 5px;

        &:hover {
          color: #eee;
        }
      }
    }

    .footer {
      border-top: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;
      padding: $unnnic-spacing-inset-xs;

      button {
        width: 100%;
      }
    }

    opacity: 0;
    border-radius: 4px;
    overflow: hidden;
    position: relative;
    transition: right 0.2s ease-out,
    opacity 0.5s;
  }
}