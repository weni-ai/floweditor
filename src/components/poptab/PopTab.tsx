import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { applyVueInReact } from 'veaury';

import styles from './PopTab.module.scss';

// @ts-ignore
import Unnnic from '@weni/unnnic-system';

export interface PopTabProps {
  type?: string;
  color: string;
  bgColor?: string;
  icon: string;
  label: string;
  header: string;
  top: string;
  visible: boolean;
  onShow: () => void;
  onHide: () => void;
}

export interface PopTabState {
  width: number;
}

const UnnnicIcon = applyVueInReact(Unnnic.unnnicIcon);

export class PopTab extends React.Component<PopTabProps, PopTabState> {
  private poppedEle: HTMLDivElement;

  constructor(props: PopTabProps) {
    super(props);
    this.state = { width: 250 };

    bindCallbacks(this, {
      include: [/^handle/],
    });
  }

  public componentDidUpdate(): void {
    if (this.state.width !== this.poppedEle.offsetWidth) {
      this.setState({ width: this.poppedEle.offsetWidth });
    }
  }

  private handlePoppedRef(ref: HTMLDivElement): HTMLDivElement {
    return (this.poppedEle = ref);
  }

  private handleTabClick(): void {
    this.props.onShow();
  }

  private handleClose(): void {
    this.props.onHide();
  }
  public render(): JSX.Element {
    return (
      <div
        className={
          'pop_wrapper ' +
          styles.pop_wrapper +
          ' ' +
          (this.props.visible ? styles.visible : '')
        }
        style={{
          right: -this.state.width,
          top: this.props.top,
        }}
      >
        <div className={styles.tab_wrapper}>
          <div
            className={styles.tab}
            style={{ background: this.props.color }}
            onClick={this.handleTabClick}
          >
            <div className={styles.icon}>
              <UnnnicIcon
                icon={this.props.icon}
                size="lg"
                scheme="neutral-snow"
              />
            </div>
            <div className={styles.label}>{this.props.label}</div>
          </div>
        </div>
        <div
          ref={this.handlePoppedRef}
          className={`${styles.popped} ${styles[this.props.type]}`}
          style={{
            borderColor: this.props.type !== 'revisions' && this.props.color,
            right: this.props.visible ? 15 + this.state.width : 0,
            top: -100,
          }}
        >
          <div
            className={styles.header}
            style={{ background: this.props.color }}
          >
            <div className={styles.close} onClick={this.handleClose}>
              <UnnnicIcon icon={'close'} size="md" scheme="neutral-snow" />
            </div>
          </div>
          <div
            className={styles.body}
            style={{ background: this.props.bgColor || this.props.color }}
          >
            {this.props.children}
          </div>
          {this.props.type === 'revisions' && (
            <div
              className={styles.footer}
              style={{ background: this.props.color }}
            />
          )}
        </div>
      </div>
    );
  }
}
