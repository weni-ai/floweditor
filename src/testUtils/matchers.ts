import { commaListsOr } from 'common-tags';
import { Exit } from 'flowTypes';
import { RenderNode } from 'store/flowContext';

const snapshot = require('jest-snapshot');

const matchers: jest.ExpectExtendMap = {};

interface MatchResult {
  message: () => string;
  pass: boolean;
}

function toBeUnique<T>(this: jest.MatcherUtils, received: any[]): MatchResult {
  const seen: { [value: string]: boolean } = {};

  for (const item of received) {
    const stringified = JSON.stringify(item, null, 1);
    if (seen[stringified]) {
      return {
        message: () => `Duplicate item in array:\n${stringified}`,
        pass: false,
      };
    } else {
      seen[stringified] = true;
    }
  }

  return {
    message: () => 'Array is unique',
    pass: true,
  };
}

function toMatchCallSnapshot<T>(
  this: jest.MatcherUtils,
  received: any,
  snapshotName?: string,
): MatchResult {
  return snapshot.toMatchSnapshot.call(
    this,
    received.mock.calls[0],
    snapshotName,
  );
}

function toHaveInboundConnections<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
): MatchResult {
  if (Object.keys(received.inboundConnections).length > 1) {
    return {
      message: () => `Node ${received.node.uuid} has an inbound connection`,
      pass: true,
    };
  }

  return {
    message: () => `Node ${received.node.uuid} has no inbound connections`,
    pass: false,
  };
}

function toHaveInboundFrom<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
  expected: Exit,
): MatchResult {
  for (const exitUUID of Object.keys(received.inboundConnections)) {
    if (exitUUID === expected.uuid) {
      return {
        message: () =>
          `Node ${received.node.uuid} incorrectly has inbound from ${expected.uuid}`,
        pass: true,
      };
    }
  }

  return {
    message: () =>
      `Node ${received.node.uuid} doesn't have inbound connection from ${expected.uuid}`,
    pass: false,
  };
}

function toPointTo<T>(
  this: jest.MatcherUtils,
  received: Exit,
  expected: RenderNode,
): MatchResult {
  if (received.destination_uuid === expected.node.uuid) {
    return {
      message: () =>
        `${received.uuid} incorrectly points to ${expected.node.uuid}`,
      pass: true,
    };
  }

  return {
    message: () =>
      `Expected ${received.uuid} to point to ${expected.node.uuid}`,
    pass: false,
  };
}

function toHaveExitWithDestination<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
): MatchResult {
  for (const exit of received.node.exits) {
    if (exit.destination_uuid !== null) {
      return {
        message: () =>
          `Exit ${exit.uuid} in node ${received.node.uuid} has a destination`,
        pass: true,
      };
    }
  }

  return {
    message: () => `All exits for ${received.node.uuid} are disconnected`,
    pass: false,
  };
}

function toHaveExitThatPointsTo<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
  expected: RenderNode,
): MatchResult {
  for (const exit of received.node.exits) {
    if (exit.destination_uuid === expected.node.uuid) {
      return {
        message: () =>
          `${exit.uuid} incorrectly has exit that points to ${expected.node.uuid}`,
        pass: true,
      };
    }
  }

  return {
    message: () =>
      `Expected ${received.node.uuid} to have an exit that points to ${expected.node.uuid}`,
    pass: false,
  };
}

function toHavePayload<T>(
  this: jest.MatcherUtils,
  store: any,
  actionType: string,
  expectedPayload: any,
): MatchResult {
  const payload = JSON.stringify(expectedPayload);
  for (const actionTaken of store.getActions()) {
    if (actionTaken.type === actionType) {
      if (JSON.stringify(actionTaken.payload) === payload) {
        return {
          message: () =>
            `Result contained action type ${actionType} with payload ${payload}`,
          pass: true,
        };
      }
    }
  }

  return {
    message: () =>
      `Could not find action type ${actionType} with payload ${payload}`,
    pass: false,
  };
}

function toHaveReduxActions<T>(
  this: jest.MatcherUtils,
  store: any,
  actionTypes: string[],
): MatchResult {
  const missedTypes: any[] = [];
  for (const actionTaken of store.getActions()) {
    for (const actionType of actionTypes) {
      if (actionTaken.type === actionType) {
        return {
          message: () => `Result contained action type ${actionType}`,
          pass: true,
        };
      } else {
        missedTypes.push(actionType);
      }
    }
  }

  return {
    message: () =>
      commaListsOr`Could not find action type(s): ${[...missedTypes]}`,
    pass: false,
  };
}

expect.extend({
  toPointTo,
  toHaveExitThatPointsTo,
  toHaveInboundFrom,
  toHaveExitWithDestination,
  toHaveInboundConnections,
  toHavePayload,
  toHaveReduxActions,
  toMatchCallSnapshot,
  toBeUnique,
});
