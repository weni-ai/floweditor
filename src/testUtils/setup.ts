/* eslint-disable no-redeclare */
import { Console } from 'console';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Exit } from 'flowTypes';
import { RenderNode } from 'store/flowContext';

// Declare custom matcher types
declare global {
  namespace jest {
    interface Matchers<R> {
      toPointTo(renderNode: RenderNode): R;
      toHaveExitThatPointsTo(renderNode: RenderNode): R;
      toHaveInboundFrom(exit: Exit): R;
      toHaveExitWithDestination(): R;
      toHaveInboundConnections(): R;
      toHavePayload(action: string, payload: any): R;
      toHaveReduxActions(actions: string[]): R;
      toMatchCallSnapshot(snapshotName?: string): R;
    }
  }
}

// Ensure console logs are visible while running tests https://github.com/facebook/jest/issues/3853
global.console = new Console(process.stderr, process.stderr);

// Configure Enzyme adapter
configure({ adapter: new Adapter() });

// RAF shim
// tslint:disable-next-line:ban-types
(global as any).requestAnimationFrame = (callback: Function) => {
  setTimeout(callback, 0);
};
