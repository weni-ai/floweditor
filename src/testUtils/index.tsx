// TODO: Remove use of Function
// tslint:disable:ban-types
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { FlowDefinition, FlowEditorConfig } from 'flowTypes';
import mutate, { Query } from 'immutability-helper';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import createStore from 'store/createStore';
import { RenderNodeMap } from 'store/flowContext';
import { getFlowComponents } from 'store/helpers';
import AppState, { initialState } from 'store/state';
import config from 'test/config';
import * as matchers from 'testUtils/matchers';
import { merge, set } from 'utils';

// we need to use require syntax to bust implicit any
const boring: FlowDefinition = require('test/flows/boring.json');

// force our matchers to be read in
const match = matchers;

export interface Resp {
  assets: { [key: string]: any }[];
}

export interface QueryString {
  [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
  config: fakePropType,
  store: fakePropType,
  assetService: fakePropType,
};

export const baseState: AppState = mutate(initialState, {
  flowContext: merge({
    definition: require('test/flows/colors.json') as FlowDefinition,
  }),
  editorState: merge({
    language: { id: 'eng', name: 'English' },
  }),
});

const flowEditorConfig: FlowEditorConfig = config;

export const configProviderContext: ConfigProviderContext = {
  config: flowEditorConfig,
};

export const setMock = (
  implementation?: (...args: any[]) => any,
): Query<jest.Mock> => set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
  Component: React.ComponentClass | React.SFC,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => (
  shallowRender: boolean = true,
  propOverrides: Query<P | Partial<P>> = {},
  duxStateOverrides: Query<AppState | Partial<AppState>> = {},
  contextOverrides: Query<
    ConfigProviderContext | Partial<ConfigProviderContext>
  > = {},
  childContextTypeOverrides: { [key: string]: Function } = {},
  children: (JSX.Element | React.ComponentClass | React.SFC)[] = [],
) => {
  const props = mutate(baseProps, propOverrides);
  const store = createStore(mutate(
    baseDuxState,
    duxStateOverrides,
  ) as AppState);
  let context = mutate(baseContext, merge({ store }));

  if (Object.keys(duxStateOverrides).length > 0) {
    context = mutate(context, duxStateOverrides);
  }

  if (Object.keys(contextOverrides).length > 0) {
    context = mutate(context, contextOverrides as ConfigProviderContext);
  }

  const childContextTypes: { [contextProp: string]: Function } = mutate(
    contextTypes,
    childContextTypeOverrides,
  );
  // tslint:disable-next-line:ban-types
  const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
    <Component {...props}>{children}</Component>,
    {
      context,
      childContextTypes,
    },
  );

  return {
    wrapper,
    props,
    context,
    instance: wrapper.instance(),
  };
};

export const composeSpy = (obj: Record<string, any> | React.ComponentClass) => (
  instanceMethod: string,
) =>
  jest.spyOn(
    (obj as React.ComponentClass).prototype || obj,
    instanceMethod as any,
  );

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () =>
  new Promise((resolve: any) => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance<any>[]) =>
  spies.forEach((spy: any) => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
  componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
  specName: string,
  attributeName: string = 'data-spec',
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
  query: Query<AppState | Partial<AppState>> = {},
  duxState = baseState,
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
  Component: React.ComponentClass | React.SFC | any,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => ({
  setup: composeSetup<P>(Component, baseProps, baseDuxState, baseContext),
  spyOn: composeSpy(Component),
});

export const prepMockDuxState = (): {
  testNodes: RenderNodeMap;
  mockDuxState: AppState;
} => {
  const testNodes = getFlowComponents(boring).renderNodeMap;

  return {
    testNodes,
    mockDuxState: {
      ...initialState,
      flowContext: {
        ...initialState.flowContext,
        definition: boring,
        nodes: testNodes,
      },
    },
  };
};

export const createMockStore: Function = configureStore([thunk]);

export const mock = <T extends {}, K extends keyof T>(
  object: T,
  property: K,
  value: T[K],
) => {
  Object.defineProperty(object, property, { get: () => value });
};
