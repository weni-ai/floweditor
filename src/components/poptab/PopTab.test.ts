import { PopTab, PopTabProps } from 'components/poptab/PopTab';
import { composeComponentTestUtils } from 'testUtils';

const baseProps: PopTabProps = {
  visible: true,
  label: 'Show Revisions',
  top: '50px',
  color: 'red',
  icon: 'fe-history',
  header: 'Revisions',
  onShow: jest.fn(),
  onHide: jest.fn(),
};

const { setup } = composeComponentTestUtils(PopTab, baseProps);

describe(PopTab.name, () => {
  describe('render', () => {
    it('should render base component', async () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
