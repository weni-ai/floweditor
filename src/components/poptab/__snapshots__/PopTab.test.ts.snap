// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopTab > render > should render base component 1`] = `
<div
  className="pop_wrapper _pop_wrapper_6f73c3 _visible_6f73c3"
  style={
    {
      "right": -250,
      "top": "50px",
    }
  }
>
  <div
    className="_tab_wrapper_6f73c3"
  >
    <div
      className="_tab_6f73c3"
      onClick={[Function]}
      style={
        {
          "background": "red",
        }
      }
    >
      <div
        className="_icon_6f73c3"
      >
        <ForwardRef
          icon="fe-history"
          scheme="neutral-snow"
          size="lg"
        />
      </div>
      <div
        className="_label_6f73c3"
      >
        Show Revisions
      </div>
    </div>
  </div>
  <div
    className="_popped_6f73c3 _undefined_6f73c3"
    style={
      {
        "borderColor": "red",
        "right": 265,
        "top": -100,
      }
    }
  >
    <div
      className="_header_6f73c3"
      style={
        {
          "background": "red",
        }
      }
    >
      <div
        className="_close_6f73c3"
        onClick={[Function]}
      >
        <ForwardRef
          icon="close"
          scheme="neutral-snow"
          size="md"
        />
      </div>
    </div>
    <div
      className="_body_6f73c3"
      style={
        {
          "background": "red",
        }
      }
    />
  </div>
</div>
`;
